test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/regex/RegexMatcherTest.scala

```diff
@@ -33,42 +33,129 @@ class RegexMatcherTest  {
-    checkGroup0(matcher.start(0), matcher.end(0), matcher.group(0))
-    checkGroup1(matcher.start(1), matcher.end(1), matcher.group(1))
-    checkGroup2(matcher.start(2), matcher.end(2), matcher.group(2))
-    checkGroup3(matcher.start(3), matcher.end(3), matcher.group(3))
+    checkGroups(matcher,
+      (0,  8,  "Scala.js"),
+      (12, 16, "nice"),
+      (0,  5,  "Scala")
+    )
+    assertTrue(matcher.find())
+    checkGroups(matcher,
+      (18, 26, "Scala.js"),
+      (30, 37, "awesome"),
+      (18, 23, "Scala")
+    )
+  }
 
+  def parseExpect(regex: String, str: String, pos: Int*): Unit = {
+    val matcher = Pattern.compile(regex).matcher(str)
+    assertTrue(matcher.find())
+    assertEquals(pos.length / 2 - 1, matcher.groupCount)
+    for (i <- 0 until pos.length by 2) {
+      assertEquals(pos(i), matcher.start(i/2))
+      assertEquals(pos(i+1), matcher.end(i/2))
+    }
+  }
+
+  @Test def parseRegex_test(): Unit = {
+    parseExpect("aa", "aa",         0, 2)
+    parseExpect("a(a)", "aa",       0, 2,  1, 2)
+    parseExpect("ABC(A(B))(C)",  "ABCABC", 0, 6,  3, 5,  4, 5,  5, 6)
+    parseExpect("A(?:A)", "AA",     0, 2)
+    parseExpect("A(?:(\\d))", "A1", 0, 2,  1, 2)
+    parseExpect("A((?:A))", "AA",   0, 2,  1, 2)
+    parseExpect("ab((ab))", "abab", 0, 4,  2, 4,  2, 4)
+    parseExpect("hum(hum)?", "humhum",     0, 6,  3, 6)
+    parseExpect("hum(hum)?", "hum",        0, 3, -1,-1)
+    parseExpect("hum(?=hum)", "humhum",    0, 3)
+    parseExpect("hum(?!hum)", "humhumhuf",        3, 6)
+    parseExpect("hum(?=h(um))", "humhum",  0, 3,  4, 6)
+    parseExpect("abab(ab){1,2}", "abababab",      0, 8,  6, 8)
+    parseExpect("abab(ab){1,2}(abc){1,2}", "abababababcabc",     0, 14,  6, 8,  11, 14)
+    parseExpect("ab(ab)?ab(?=aba)(ab)*", "abababababa",  0, 10,  2, 4,   8, 10)
+    parseExpect("ab(?=aba)ab(aba)?(ab)*", "abababababa", 0, 7,   4, 7,  -1, -1)
+    parseExpect("ab(?=aba)ab(aba)??(ab)*","abababababa", 0, 10, -1, -1,  8, 10)
+    parseExpect("abab(ab){1,2}?", "abababab",     0, 6,  4, 6)
+    parseExpect("ab(?:ab)*", "abababab",   0, 8)
+    if (!executingInJVM) {
+      parseExpect("ab(?:a(c))*ac", "abacacac",      0, 8,  5, 6)
+      parseExpect("ab(?:a(c))+ac", "abacacac",      0, 8,  5, 6)
+    }
+    parseExpect("ab(?:ac)*?ac", "abacacac",       0, 4)
+    parseExpect("ab(?:ac)+?ac", "abacacac",       0, 6)
+    parseExpect("ab((?=abab(ab))a(b))*a", "abababab",    0, 5,  2, 4,  6, 8,  3, 4)
+  }
+
+  @Test def parseRegex_backgroups_test(): Unit = {
+    parseExpect("bc(.c).c(\\1)", "bczcxczc",    0, 8,  2, 4,  6, 8)
+    parseExpect("(bc(.c).c)(\\2)", "bczcxczc",  0, 8,  0, 6,  2, 4,  6, 8)
+  }
+
+  def checkGroups(matcher: Matcher, startEndMatch: (Int, Int, String)*): Unit = {
+    assertEquals(startEndMatch(0)._1, matcher.start)
+    assertEquals(startEndMatch(0)._2, matcher.end  )
+    assertEquals(startEndMatch(0)._3, matcher.group)
+    for (((start, end, mtch), i) <- startEndMatch.zipWithIndex) {
```

Comment:
You should add
```scala
assertEquals(startEndMatch.size - 1, matcher.groupCount)
```
above this line, and then you can get rid of the `assertEquals(N, matcher.groupCount)` that are in all the specific test cases.

Unless I'm missing something ...

Reply:
Did so. Also, I factored the `assertTrue(matcher.find())` in `checkGroups` since they are always called before.